### PR TITLE
[Rust] Second stage of Rust Refactor

### PR DESCRIPTION
**Pull request by jroesch**
See #5526, this PR implements the object system and basic types.

### CODE REVIEW

File: rust/tvm-rt/README.md [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+<!--- Unless required by applicable law or agreed to in writing, -->
+<!--- software distributed under the License is distributed on an -->
+<!--- "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY -->
+<!--- KIND, either express or implied.  See the License for the -->
+<!--- specific language governing permissions and limitations -->
+<!--- under the License. -->
+
+# TVM Runtime Frontend Support
+
+This crate provides an idiomatic Rust API for [TVM](https://github.com/apache/incubator-tvm) runtime frontend. Currently this requires **Nightly Rust** and tested on `rustc 1.32.0-nightly`
```
Comment (ehsanmok) [inline rust/tvm-rt/README.md:20]: Nudging to update the READMEs when done.

Comment (jroesch) [reply rust/tvm-rt/README.md:20]: yeah totally on my radar :) plan is to rebase this once I land the other changes, then a final patch to add the compiler part and stitch it all together. I'll then update the graph runtime. Finally I'll probably send a 4th PR flipping CI on for the new bindings and removing the old ones.

**Comment by leonwanghui** [comment]
@jroesch It seems quite promising for Rust bindings, so any updates on this PR?

**Comment by jroesch** [comment]
@leonwanghui I have like 3-4 different branches that I sliced my like 6-8kloc changes. This one is now ready for review, I would like to land this, then land the frontend refactor, then switch over the runtime and CI to the new bindings.

File: include/tvm/runtime/c_runtime_api.h [hunk 2]
```diff
@@ -514,16 +523,6 @@ TVM_DLL int TVMObjectTypeKey2Index(const char* type_key, unsigned* out_tindex);
  */
 TVM_DLL int TVMObjectFree(TVMObjectHandle obj);
 
-/*!
```
Comment (jroesch) [inline include/tvm/runtime/c_runtime_api.h:517]: Restore.

File: rust/macros/src/external.rs [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        }
+
+        Ok(ExternalInput { externs })
+    }
+}
+
+pub fn macro_impl(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let ext_input = syn::parse_macro_input!(input as ExternalInput);
+
+    let tvm_rt_crate = if env::var("CARGO_PKG_NAME").unwrap() == "tvm-rt" {
```
Comment (jroesch) [inline rust/macros/src/external.rs:90]: Refactor to use helper.

File: rust/tvm-rt/README.md (hunk 1 above)
Comment (jroesch) [reply rust/tvm-rt/README.md:20]: Most of these docs will live in the `tvm` crate. I cleaned up the current README.md and will add more back as the bindings stablize.

File: rust/tvm-rt/src/function.rs [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+
+        function::register_override(constfn, "constfn".to_owned(), true).unwrap();
+        let func = Function::get("constfn").unwrap();
+        let func = func.to_boxed_fn::<dyn Fn() -> Result<i32>>();
+        let ret = func().unwrap();
+        assert_eq!(ret, 10);
+    }
+
+    // #[test]
+    // fn register_and_call_closure1() {
```
Comment (jroesch) [inline rust/tvm-rt/src/function.rs:333]: Turn back on.

File: rust/tvm-rt/src/object/object_ptr.rs [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+    }
+}
+
+pub unsafe trait IsObject {
+    const TYPE_KEY: &'static str;
+
+    fn as_object<'s>(&'s self) -> &'s Object;
+
+    unsafe extern "C" fn typed_delete(_object: *mut Self) {
+        // let object = Box::from_raw(object);
```
Comment (jroesch) [inline rust/tvm-rt/src/object/object_ptr.rs:101]: Restore deallocation code.

File: rust/tvm-rt/src/object/object_ptr.rs [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+impl<T> Clone for ObjectPtr<T> {
+    fn clone(&self) -> Self {
+        unsafe {
+            let raw_ptr = std::mem::transmute(self.ptr);
+            assert_eq!(TVMObjectRetain(raw_ptr), 0);
+            ObjectPtr { ptr: self.ptr }
+        }
+    }
+}
+
```
Comment (jroesch) [inline rust/tvm-rt/src/object/object_ptr.rs:135]: Restore drop impl.

File: rust/tvm-rt/src/to_boxed_fn.rs [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+//! This module provides an idiomatic Rust API for creating and working with TVM functions.
+//!
+//! For calling an already registered TVM function use [`function::Builder`]
```
Comment (jroesch) [inline rust/tvm-rt/src/to_boxed_fn.rs:22]: Update comment.

File: rust/tvm-rt/README.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+```rust
+use tvm_rt::{ArgValue, RetValue};
+use tvm_rt::function::{Function, Result, register};
+
+fn sum(x: i64, y: i64, z: i64) -> i64 {
+    x + y + z
+}
+
+fn main() {
+    register(sum, "mysum".to_owned()).unwrap();
```
Comment (leonwanghui) [inline rust/tvm-rt/README.md:54]: `register` or `register!`?

**Comment by leonwanghui** [comment]
@jroesch I think it's cool to have these crates refactored, but before they are all landed, would you mind illustrating the final landscape of these crates? Thanks!

File: rust/macros/src/lib.rs [hunk 9]
```diff
@@ -17,121 +17,25 @@
...
 }
 
-impl Parse for ImportModule {
-    fn parse(input: ParseStream) -> Result<Self> {
-        let importing_file: LitStr = input.parse()?;
-        Ok(ImportModule { importing_file })
-    }
+#[proc_macro_derive(Object, attributes(base, ref_name, type_key))]
+pub fn macro_impl(input: TokenStream) -> TokenStream {
+    // let input = proc_macro2::TokenStream::from(input);
```
Comment (adelbertc) [inline rust/macros/src/lib.rs:34]: delete

File: rust/tvm-rt/src/function.rs [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+        ));
+        let names_list = unsafe { slice::from_raw_parts(names_ptr, out_size as usize) };
+
+        let names_list: Vec<String> =
+            names_list
+            .iter()
+            .map(|&p| unsafe { CStr::from_ptr(p).to_str().unwrap().into() })
+            .collect();
+
+        // println!("{:?}", &names_list);
```
Comment (adelbertc) [inline rust/tvm-rt/src/function.rs:64]: remove

File: rust/tvm-rt/src/function.rs [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+mod tests {
+    use super::*;
+    use crate::function::Function;
+
+    static CANARY: &str = "runtime.ModuleLoadFromFile";
+
+    // #[test]
+    // fn list_global_func() {
+    //     assert!(GLOBAL_FUNCTIONS.lock().unwrap().contains_key(CANARY));
+    // }
```
Comment (adelbertc) [inline rust/tvm-rt/src/function.rs:308]: remove

File: rust/tvm-rt/src/object/mod.rs [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+external! {
+    #[name("ir.DebugPrint")]
+    fn debug_print(object: ObjectRef) -> CString;
+}
+
+// external! {
+//     #[name("ir.TextPrinter")]
+//     fn as_text(object: ObjectRef) -> CString;
+// }
```
Comment (adelbertc) [inline rust/tvm-rt/src/object/mod.rs:117]: remove

File: rust/tvm-rt/src/to_function.rs [hunk 13]
```diff
@@ -0,0 +1,359 @@
...
+
+    fn helper<F, I, O>(f: F) -> Function
+    where
+        F: ToFunction<I, O>,
+        F: Typed<I, O>,
+    {
+        f.to_function()
+    }
+
+    // fn func_args(args: &[ArgValue<'static>]) -> Result<RetValue> {
```
Comment (adelbertc) [inline rust/tvm-rt/src/to_function.rs:327]: remove

File: rust/tvm-rt/README.md (hunk 8 above)
Comment (robo-corg) [inline rust/tvm-rt/README.md:54]:
```suggestion
    register!(sum, "mysum".to_owned()).unwrap();
```

File: rust/tvm-rt/src/function.rs [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+where
+    F: ToFunction<I, O>,
+    F: Typed<I, O>,
+{
+    let func = f.to_function();
+    let name = name.into();
+    let mut globals = GLOBAL_FUNCTIONS.lock().unwrap();
+    // Not sure about this code
+    let handle = func.handle();
+    globals.insert(name.clone(), Some(func));
```
Comment (robo-corg) [inline rust/tvm-rt/src/function.rs:287]: What about existing &'static Function handed out? It seems like the best case is a non-mut ref is updated, worst case this reference is now pointing at something else when the BTree updates?

File: rust/tvm-rt/src/function.rs [hunk 15]
```diff
@@ -0,0 +1,345 @@
...
+            return 10;
+        }
+
+        function::register_override(constfn, "constfn".to_owned(), true).unwrap();
+        let func = Function::get("constfn").unwrap();
+        let func = func.to_boxed_fn::<dyn Fn() -> Result<i32>>();
+        let ret = func().unwrap();
+        assert_eq!(ret, 10);
+    }
+
```
Comment (robo-corg) [inline rust/tvm-rt/src/function.rs:331]:
```suggestion
    #[test]
    fn register_and_call_original_closure1() {
        use crate::function::{self};

        fn ident(x: i64) -> i64 {
            return x;
        }
        
        let func = Function::get("ident").unwrap();

        function::register_override(ident, "ident".to_owned(), false).unwrap();
        
        // What is func now?
        let func = func.to_boxed_fn::<dyn Fn(i32) -> Result<i32>>();
        assert_eq!(func(60).unwrap(), 60);
    }
```

File: rust/tvm-rt/src/to_function.rs [hunk 16]
```diff
@@ -0,0 +1,359 @@
...
+//     fn call(handle: *mut Self::Handle, args: &[ArgValue<'static>]) -> Result<RetValue> {
+//         let handle: Self::Handle = unsafe { std::mem::transmute(handle) };
+//         let r = handle(args);
+//         println!("afters");
+//         r
+//     }
+
+//     // Function's don't need de-allocation because the pointers are into the code section of memory.
+//     fn drop(_: *mut Self::Handle) {}
+// }
```
Comment (robo-corg) [inline rust/tvm-rt/src/to_function.rs:255]:
```suggestion
```
Delete commented code

File: rust/tvm-rt/src/value.rs [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+//! This module implements [`ArgValue`] and [`RetValue`] types
+//! and their conversions needed for the types used in frontend crate.
+//! `RetValue` is the owned version of `TVMPODValue`.
+
+use std::convert::TryFrom;
+// use std::ffi::c_void;
```
Comment (robo-corg) [inline rust/tvm-rt/src/value.rs:25]:
```suggestion
```
Delete

File: rust/tvm-rt/src/object/object_ptr.rs [hunk 18]
```diff
@@ -0,0 +1,306 @@
...
+use crate::errors::Error;
+
+type Deleter<T> = unsafe extern "C" fn(object: *mut T) -> ();
+
+#[derive(Debug)]
+#[repr(C)]
+pub struct Object {
+    pub type_index: u32,
+    pub ref_count: i32,
+    pub fdeleter: Deleter<Object>,
```
Comment (robo-corg) [inline rust/tvm-rt/src/object/object_ptr.rs:36]: Should these fields be pub?

File: rust/tvm-rt/src/object/object_ptr.rs [hunk 19]
```diff
@@ -0,0 +1,306 @@
...
+
+    if is_derived == 0 {
+        false
+    } else {
+        true
+    }
+}
+
+impl Object {
+    fn new(type_index: u32, deleter: Deleter<Object>) -> Object {
```
Comment (robo-corg) [inline rust/tvm-rt/src/object/object_ptr.rs:60]: Make this pub and remove pub fields from object?

File: rust/tvm-rt/src/to_function.rs [hunk 20]
```diff
@@ -0,0 +1,359 @@
...
+
+    /// The finalizer which is invoked when the packed function's
+    /// reference count is zero.
+    unsafe extern "C" fn tvm_finalizer(fhandle: *mut c_void) {
+        let handle = std::mem::transmute(fhandle);
+        Self::drop(handle)
+    }
+}
+
+// /// A wrapper that is used to work around inference issues for bare functions.
```
Comment (binarybana) [inline rust/tvm-rt/src/to_function.rs:217]: Remove?

File: rust/macros/src/import_module.rs [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+        std::env::var("CARGO_MANIFEST_DIR").expect("variable should always be set by Cargo.");
+
+    let mut path = PathBuf::new();
+    path.push(manifest);
+    path = path.join(import_module_args.importing_file.value());
+
+    let mut fd = File::open(&path)
+        .unwrap_or_else(|_| panic!("Unable to find TVM object file at `{}`", path.display()));
+    let mut buffer = Vec::new();
+    fd.read_to_end(&mut buffer).unwrap();
```
[the remark this answers is not in the document]
Comment (jroesch) [reply rust/macros/src/import_module.rs:50]: Yeah I think we should do that, I have one more architectural PR and then I'm all done.

File: rust/macros/src/object.rs [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+            }
+        }
+
+        impl std::convert::TryFrom<tvm_rt::RetValue> for #ref_id {
+            type Error = ::anyhow::Error;
+
+            fn try_from(ret_val: tvm_rt::RetValue) -> Result<#ref_id, Self::Error> {
+                use std::convert::TryInto;
+                let oref: ObjectRef = ret_val.try_into()?;
+                let ptr = oref.0.ok_or(anyhow::anyhow!("null ptr"))?;
```
[the remark this answers is not in the document]
Comment (jroesch) [reply rust/macros/src/object.rs]: I am trying to remove my over zealous use of anyhow after Jason complained.

File: rust/macros/src/external.rs [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+use proc_macro2::Span;
+use quote::quote;
+use syn::parse::{Parse, ParseStream, Result};
+
+use syn::{FnArg, Generics, Ident, Lit, Meta, NestedMeta, Pat, ReturnType, TraitItemMethod, Type};
+
+struct External {
```
[the remark this answers is not in the document]
Comment (jroesch) [reply rust/macros/src/external.rs:25]: I promise I will write a lot more docs when the refactoring is done, I've rewritten and refactored all of this so many times I want to make sure it is here to stay before I spend the cycles.

**Comment by jroesch** [comment]
@leonwanghui I have one more structural PR and when I deprecate the old crates and move to the new ones I will write more long form docs, there will be 3 crates when I'm done `tvm-sys` the low level FFI bindings, `tvm-rt` the bindings to the runtime system, and `tvm` bindings to the TVM compiler and in general a larger API surface.

File: rust/tvm-rt/src/ndarray.rs [hunk 24]
```diff
@@ -0,0 +1,439 @@
...
+    fn drop(&mut self) {
+        if let &mut NDArray::Owned { .. } = self {
+            check_call!(ffi::TVMArrayFree(self.as_raw_dltensor()));
+        }
+    }
+}
+
+mod sealed {
+    /// Private trait to prevent other traits from being implemeneted in downstream crates.
+    pub trait Sealed {}
```
[the remark this answers is not in the document]
Comment (jroesch) [reply rust/tvm-rt/src/ndarray.rs:352]: This is part of the existing code, I plan on purging in the final iteration, NDArray has changed a lot and needs to be reworked.

File: rust/tvm-rt/src/function.rs [hunk 25]
```diff
@@ -0,0 +1,345 @@
...
+        let names_list = names_list
+            .into_iter()
+            .map(|p| (p, None))
+            .collect();
+
+        Mutex::new(names_list)
+    };
+}
+
+/// Wrapper around TVM function handle which includes `is_global`
```
[the remark this answers is not in the document]
Comment (jroesch) [reply rust/tvm-rt/src/function.rs:44]: I will queue updating the docs.

**Comment by leonwanghui** [comment]
@jroesch Got it, thanks a lot for the elaboration : )

**Pull request merged**